File: MPI_Code/Fortran/gather.f90
PROGRAM gather
USE mpi_f08
IMPLICIT NONE
INTEGER::isend,irecv(4),recvdata(4),nprocs,nrank
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,nrank)
CALL MPI_Comm_size(MPI_COMM_WORLD,nprocs)
isend=nrank+1;    recvdata=0
PRINT*,'rank:',nrank,'isend:',isend
CALL MPI_Gather(isend,1,MPI_INTEGER,irecv,1,MPI_INTEGER,0,MPI_COMM_WORLD)
IF(nrank==0) PRINT*,'rank:',nrank,'irecv=',irecv
isend=isend*10
IF(nrank==nprocs-1)THEN
  recvdata(nprocs)=isend
  CALL MPI_Gather(MPI_IN_PLACE,1,MPI_INTEGER,recvdata,1,MPI_INTEGER,nprocs-1,MPI_COMM_WORLD)
ELSE
  CALL MPI_Gather(isend,1,MPI_INTEGER,recvdata,1,MPI_INTEGER,nprocs-1,MPI_COMM_WORLD)
ENDIF
IF(nrank==nprocs-1) PRINT*,'RECVDATA:',recvdata

CALL MPI_Finalize
END PROGRAM gather

File: example_code/mpi/MPI_Code/C/gather.c
#include <stdio.h>
#include "mpi.h"
int main()
{
  int isend,irecv[4]={0,},recvdata[4]={0,},nprocs,nrank;
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&nrank);
  MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
  isend=nrank+1;
  printf("rank: %d, isend:%d\n",nrank,isend);
  MPI_Gather(&isend,1,MPI_INT,irecv,1,MPI_INT,0,MPI_COMM_WORLD);
  if(nrank==0) printf("rank:%d  irecv=%d  %d  %d  %d\n",nrank,irecv[0],irecv[1],irecv[2],irecv[3]);
  isend *=10;
  if(nrank==nprocs-1){
    recvdata[nrank]=isend;
    MPI_Gather(MPI_IN_PLACE,1,MPI_INT,recvdata,1,MPI_INT,nprocs-1,MPI_COMM_WORLD);
  } else {
    MPI_Gather(&isend,1,MPI_INT,recvdata,1,MPI_INT,nprocs-1,MPI_COMM_WORLD);
  }
  if(nrank==nprocs-1) printf("RECVDATA:%d  %d  %d  %d\n",recvdata[0], recvdata[1], recvdata[2], recvdata[3]);
  MPI_Finalize();
  return 0;
}
